### PR TITLE
[New Concept] : functions

### PR DESCRIPTION
**Pull request by mukeshgurpude**
Adding functions `concept`, as discussed in #2353

### CODE REVIEW

File: concepts/functions/.meta/config.json
Comment (IsaacG) [inline concepts/functions/.meta/config.json:2]: > It can pass data, known as parameters, into a function. 

"It" here would be "functions" so this reads "functions can pass data into a function". Which is confusing and circular.

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:3]: "single" implies one. "Related" implies multiple. Related to what?

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:4]:
```suggestion
`Functions` provide better modularity for your application and a high degree of code reuse.
```

Better than what?

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:5]: These have a technical name: built-in functions. It would be good to use the proper name when they are first introduced. 
It might be good to link to them, too: `https://docs.python.org/3/library/functions.html`

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:6]:
```suggestion
You can also define your own functions. Those are called [`user-defined functions`][user defined functions].
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:7]:
```suggestion
Functions can run something as simple as _printing a message to the console_ or they can be quite complex.
```

`a + b` is a calculation and is not very complex.

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:10]:
```suggestion
Data. known as `parameters`, can be passed to the function by placing them inside the parenthesis. A broader term for parameters is `arguments`. Functions can perform different tasks depending on the value of the parameters.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:12]:
```suggestion
A function can also return a value using the `return` keyword. The value returned by the function is known as the `return value`. The return value is returned to the caller of the function.
```

Comment (IsaacG) [reply concepts/functions/about.md:12]: Has "caller" been defined?

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:16]:
```suggestion
In python, functions are created using the `def` keyword. The function definition is followed by the function name and parenthesis [`()`]. Inside the parenthesis, the parameters are specified, separated by commas. After the close parenthesis, the colon (`:`) is used to separate the function signature from the function body.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:18]:
```suggestion
The function body is a block of code that is executed when the function is called. The body of the function is indented. The indentation is important because Python relies on it to know where that block of code ends. A value can be returned from the function by using the `return` keyword, which can then be used by the caller of the function.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:35]:
```suggestion
To call a function, use the function name followed by parenthesis [`()`]. Parameters passed to the function are placed inside the parenthesis, separated by commas.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:43]:
```suggestion
The above function can be called by using the following syntax:
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:51]: This document first mentions arguments, switches to parameters and is now saying they are known as arguments. One term should be used consistently.

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:53]:
```suggestion
Let's define a function `add` which adds two numbers together:
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:59]:
```suggestion
When the function is called, the parameters are passed to the function. We need to pass values for both the parameters, otherwise a [`TypeError`][type-error] will be raised.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:70]:
```suggestion
# Passing an incorrect number of parameters will raise a `TypeError`
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:84]: You call out that return values can be of any type. What about the parameter type? That was never mentioned.

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:92]:
```suggestion
We can store the return value in a variable and then print it:
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:105]:
```suggestion
# A function without an explicit return value will return `None`
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:123]:
```suggestion
# y never gets printed, because the function exits after the return statement
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:130]:
```suggestion
Complex programs can be broken down into smaller parts. Different functions can be used to perform different specific tasks.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:163]: `nonlocal` isn't commonly used and might be best left out of this. `global` isn't that common, either.

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:166]: `outer` implies an `inner`. Outer of what? Where's the inner?

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:174]: You may want to move this line *before* the function definition so the reader knows that there's a global `x` before the function is shown.

File: concepts/functions/about.md
Comment (IsaacG) [reply concepts/functions/about.md:5]: This line also has both `python` and `Python`. Both are used throughout this file.

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:204]:
```suggestion
In python, functions can be assigned to variables and passed as arguments to other functions. They can be used as return values. Functions can also be placed into a sequence([`list`][list], [`tuple`][tuple] etc) or as value in a [`dict`][dict]. Functions can be used anywhere than any other object can be used. This is because _functions are [`first class objects`][first class objects]_.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:210]:
```suggestion
# A function can be assigned to a variable.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:211]: This looks like the real print but an alias.

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:217]:
```suggestion
# Functions can be passed as an argument to another function.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:223]:
```suggestion
# Function can be used as an item in a sequence.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:228]:
```suggestion
# Functions can be used as a value in a dictionary.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:233]:
```suggestion
# Functions can be returned from a function.
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:244]: This is not used often and is often not good practice. Maybe it should be skipped.

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:258]:
```suggestion
Functions in python have special attributes. Some of them are:
```

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:256]: This is another section that might not be well suited for teaching the concept of functions.

File: concepts/functions/about.md
Comment (IsaacG) [inline concepts/functions/about.md:288]:
```suggestion
The full list of function attributes can be found at [Python DataModel][attributes].
```

File: concepts/functions/about.md
Comment (mukeshgurpude) [reply concepts/functions/about.md:163]: Actually in learning Objective section of original issue #2353, these are mentioned.
Do you want me to remove these points?

**Comment by mukeshgurpude** [comment]
Hello @J08K, @BethanyG, @IsaacG 
Please review the PR. 
Please accept my apologies for my sloppy grammar😥.

File: concepts/functions/about.md
Comment (J08K) [reply concepts/functions/about.md:163]: It is in the learning objectives, so keep them. Even though they aren't "used often" that doesn't mean people shouldn't know what they are.

File: concepts/functions/about.md
Comment (J08K) [reply concepts/functions/about.md:245]: Please read the connected issue before reviewing. `nested functions` is in the concepts of #2353. Keep nested functions in the exercise.

Nested functions are not bad practice at all. They can be used to access the enclosing namespace. If such function is only used by one other function, then you could opt in to use `nested` or `inner functions`.

Comment (J08K) [reply concepts/functions/about.md:245]: If you really would like these gone. Throw up an issue on the repo. We could either then put this issue on hold, or merge this issue (if everything else is okay) and then change the docs afterwards.

File: concepts/functions/about.md
Comment (J08K) [reply concepts/functions/about.md:257]: Please reply to the connected issue with your concerns, since it is inside of the concepts. This is not the place for such discussions.

File: concepts/functions/about.md
Comment (BethanyG) [reply concepts/functions/about.md:245]: I think [issue 2353](https://github.com/exercism/python/issues/2353) is pretty clear. The intent of this exercise is to dig deeper into Python functions - both as a "next step" for students and as an introduction into `higher order functions` and FP.  As well as more practice before students are "hit" with all the function parameter options, and unpacking actions (function arguments exercise).

Python's object model (as opposed to Javas object model, for example) treats `function objects` as first-class citizens.  It is possible (perhaps not _advisable_) to write nested functions -- or even constructs  where one function encloses multiple other functions. Students need to know that -- even if they don't end up using it all that often.

And that also helps teach scope (along with locals and globals) -- which (we hope) leads eventually to understanding what a `closure` is, and why you'd want one, and other things like `decorators`.

Comment (BethanyG) [reply concepts/functions/about.md:245]: This was not intended as a "basic" exercise.  We have `lasgana` -- and all the exercises in-between for that.  Students would not be taking on this exercise until after they finished all the exercises on data structures and sequences and more.

**Comment by BethanyG** [comment]
@mukeshgurpude  -- we all have sloppy grammar (_I can be a big offender there_), so there is no need to apologize!  It's why we have reviews. 😄   🌈 

Although I tend to want to do content review passes before grammar passes -- since re-working content often leads to additional typos or potential issues.  

But don't worry - and don't get intimidated.  We're all here to work on this and make it the best it can be.  And _**thank you**_ for submitting!  I probably won't have time to give a thorough review until tomorrow evening my time (I'm in PST) - but I'll have some more comments then. 🌟

**Comment by J08K** [comment]
So, what @BethanyG said. I agree, we might need to _hint_ a little towards the more _advanced_ uses of functions. It is ok (in my opinion) if you don't cover these in full detail, but we need to let students know that there is more than just the `def` keyword. 

I really like everything you've written so far, and I'd love to hear about what you have to say.

**Comment by BethanyG** [comment]
@mukeshgurpude   In the spirit of "optimistic merging", I am going to approve this PR.  We can continue to discuss and change it in follow-on requests.  Apologies for letting it "fall through the cracks" -- I do think that we should discuss the "advanced" part of this and map out in more detail what we intend for functions.  But rather than drag this particular PR on, I think we're better served with discussion via an issue, and then making follow-on PRs.

**Review by BethanyG — APPROVED** [review]

**Pull request merged**
